roachtest: add node-kill operation

itsbilal opened the pull request:
This change adds a node-kill operation with 4 variants: one that drains and one that doesn't x {SIGKILL, SIGTERM}.

Epic: none

Release note: None

Commit:
roachtest: add node-kill operation

This change adds a node-kill operation with 4 variants:
one that drains and one that doesn't x {SIGKILL, SIGTERM}.

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/122899)
<!-- Reviewable:end -->

herkolategan commented on pkg/cmd/roachtest/operations/node_kill.go:
```diff
@@ -0,0 +1,122 @@
+		if err != nil {
+			o.Fatal(err)
+		}
+		err = c.RunE(ctx, option.WithNodes(c.Node(nid)),
```
@renatolabs may know how to structure this better, but we could instead of doing an `ExternalPGUrl` use the new command util to optionally add `insecure` or `certs` to `drain`:

Something like this (untested):
```
		cmd := roachtestutil.NewCommand(fmt.Sprintf("%s node drain", test.DefaultCockroachPath)).
				Flag("logtostderr", "INFO").
				Flag("port", fmt.Sprintf("{pgport:%d}", nid)).
				MaybeOption(!c.IsSecure(),"insecure").
				MaybeFlag(c.IsSecure(),"certs", install.CockroachNodeCertsDir).
				Option("self")
		err = c.RunE(ctx, option.WithNodes(c.Node(nid)), cmd.String())
```

herkolategan reviewed (commented):
Left a few small comments, mostly `roachtest` convenience APIs related. One general question, It seems that clean-up is responsible for the timing regarding when the node is started back up again. Would there be value in having nodes be down for a specified (random) amount of time, rather than a fixed amount of time?

renatolabs commented on pkg/cmd/roachtest/operations/node_kill.go:
```diff
@@ -0,0 +1,122 @@
+	// We might need to restart the node if it isn't live.
+	db, err := c.ConnE(ctx, o.L(), cl.nodes[0])
+	if err != nil {
+		c.Run(ctx, option.WithNodes(cl.nodes), "./cockroach.sh")
```
Should we log this error for understandability? Same for the error below.

renatolabs commented on pkg/cmd/roachtest/operations/node_kill.go:
```diff
@@ -0,0 +1,122 @@
+
+	stopOpts := option.DefaultStopOpts()
+	stopOpts.RoachprodOpts.Sig = signal
+	if err := c.StopE(ctx, o.L(), stopOpts, c.Node(nid)); err != nil {
```
`c.Stop`?

Also, in the context of operations, this might have undesired side effects. Not sure if operations are running concurrently at the moment, but `Stop*` will stop _every_ command started by roachprod. So if another operation is running something else on the same node, that will be killed as well.

itsbilal replied:
That's a good point about concurrent commands being run using roachprod. I switched to a pkill-based approach that should be safer, hope that works better!

renatolabs commented on pkg/cmd/roachtest/operations/node_kill.go:
```diff
@@ -0,0 +1,122 @@
+	o.Status(fmt.Sprintf("killing node %d with signal %d", nid, signal))
+
+	stopOpts := option.DefaultStopOpts()
+	stopOpts.RoachprodOpts.Sig = signal
```
Do we want to `Wait`?

itsbilal replied:
It's probably unnecessary for sigkill, but yeah for sigterm it might make sense. Added.

renatolabs commented on pkg/cmd/roachtest/operations/node_kill.go:
```diff
@@ -0,0 +1,122 @@
+
+		url, err := c.ExternalPGUrl(ctx, o.L(), c.Node(nid), roachprod.PGURLOptions{
+			External: false,
+			Secure:   c.IsSecure(),
```
How does this work work? How does it know the cluster is secure a corresponding `Start` call?

itsbilal replied:
`runOperation` takes care of passing in the certs directory if that was passed to `run-operation`. And we avoid calling `Start` in operations so we don't mess up any previously-used start flags, instead preferring to call `./cockroach.sh`. Does that sound good?

renatolabs commented on pkg/cmd/roachtest/operations/node_kill.go:
```diff
@@ -0,0 +1,122 @@
+	r.AddOperation(registry.OperationSpec{
+		Name:             "node-kill/sigterm/drain=false",
+		Owner:            registry.OwnerServer,
+		Timeout:          10 * time.Minute,
```
Should we have a shorter timeout for drain=false variants?

itsbilal replied:
Good point

renatolabs commented on pkg/cmd/roachtest/operations/node_kill.go:
```diff
@@ -0,0 +1,122 @@
+	})
+	r.AddOperation(registry.OperationSpec{
+		Name:             "node-kill/sigterm/drain=false",
+		Owner:            registry.OwnerServer,
```
Not sure we should be assigning anything to the server "team"? 😅

itsbilal replied:
Haha, good point. I copied this from one of the roachtests. Any idea on what other team can own this? (We don't auto-create issues for operations so this means little other than for our own tracking).

renatolabs replied:
> I copied this from one of the roachtests

Which one? A basic search doesn't reveal any roachtests owned by server, but maybe ownership is assigned in some indirect way.

> Any idea on what other team can own this?

Maybe server is fine. I guess eventually these issue will be owned by a "DRT" team that will have to investigate and triage appropriately.

itsbilal replied:
> Which one? A basic search doesn't reveal any roachtests owned by server, but maybe ownership is assigned in some indirect way.

Oops, yeah I wasn't searching right. And yeah we can keep it assigned to server for now, with the understanding that all operation failures get triaged by DRT anyway.

itsbilal commented:
Thanks for the review! Addressed the points brought up.

> Would there be value in having nodes be down for a specified (random) amount of time, rather than a fixed amount of time?

This is a good point, and something I'd want `drt-run` (which I have been working on again lately) to handle. Currently `run-operation --wait-before-cleanup <duration>` configures it. Because yes, we'd ideally want a more interesting interleaving of operation-induced states.

renatolabs commented on pkg/cmd/roachtest/operations/node_kill.go:
```diff
@@ -0,0 +1,139 @@
+	}
+	o.Status(fmt.Sprintf("killing node %s with signal %d", node.NodeIDsString(), signal))
+
+	err := c.RunE(ctx, option.WithNodes(node), "pkill", fmt.Sprintf("-%d", signal), "-f", "cockroach\\ start")
```
I haven't tested this but I believe the following would work and would allow us to use the existing wait implementation within roachprod:

```go
stopOpts := option.StopVirtualClusterOpts(install.SystemInterfaceName, node)
stopOpts.RoachprodOpts.Wait = true
stopOpts.RoachprodOpts.MaxWait = 600 // 10 minutes
c.StopServiceForVirtualCluster(ctx, o.L(), stopOpts)
```

This will match only the `system` cockroach process started by roachprod.

itsbilal replied:
Tested this, it works well. Thanks!

itsbilal commented:
TFTR!

bors r=renatolabs

yuzefovich commented:
The linter is not happy.

bors r-

itsbilal commented:
Apologies, didn't realize the linter isn't under `Bazel essential CI` anymore.

itsbilal commented:
Required checks are green now.

bors r=renatolabs
